firebase-firestore/src/main/java/com/google/firebase/firestore/model/FieldIndex.java

```diff
@@ -94,19 +95,67 @@ public int compareTo(Segment other) {
   /** Stores the "high water mark" that indicates how updated the Index is for the current user. */
   @AutoValue
   public abstract static class IndexState {
-    public static IndexState create(long sequenceNumber, SnapshotVersion readTime) {
-      return new AutoValue_FieldIndex_IndexState(sequenceNumber, readTime);
+    public static IndexState create(long sequenceNumber, IndexOffset offset) {
+      return new AutoValue_FieldIndex_IndexState(sequenceNumber, offset);
+    }
+
+    public static IndexState create(
+        long sequenceNumber, SnapshotVersion readTime, DocumentKey documentKey) {
+      return create(sequenceNumber, IndexOffset.create(readTime, documentKey));
     }
 
     /**
      * Returns a number that indicates when the index was last updated (relative to other indexes).
      */
     public abstract long getSequenceNumber();
 
+    /** Returns the latest indexed read time and document. */
+    public abstract IndexOffset getOffset();
+  }
+
+  /** Stores the latest read time and document that were processed for an index. */
+  @AutoValue
+  public abstract static class IndexOffset implements Comparable<IndexOffset> {
+    public static final IndexOffset NONE = create(SnapshotVersion.NONE, DocumentKey.empty());
+
+    /**
+     * Creates an offset that matches all documents with a read time higher than {@code readTime} or
+     * with a key higher than {@code documentKey} for equal read times.
+     */
+    public static IndexOffset create(SnapshotVersion readTime, DocumentKey documentKey) {
+      return new AutoValue_FieldIndex_IndexOffset(readTime, documentKey);
+    }
+
+    /**
+     * Creates an offset that matches all documents with a read time higher than {@code readTime}.
+     */
+    public static IndexOffset create(SnapshotVersion readTime) {
+      long successorSeconds = readTime.getTimestamp().getSeconds();
+      int successorNanos = readTime.getTimestamp().getNanoseconds() + 1;
```

Comment:
Added a comment:

```
      // We want to create an offset that matches all documents with a read time greater than
      // the provided read time. To do so, we technically need to create an offset for
      // `(readTime, MAX_DOCUMENT_KEY)`. While we could use Unicode codepoints to generate
      // MAX_DOCUMENT_KEY, it is much easier to use `(readTime + 1, DocumentKey.empty())` since
      // `> DocumentKey.empty()` matches all valid document IDs.
```